File: src/main/java/dev/jolvera/finjav/controllers/AccountDialogController.java
package dev.jolvera.finjav.controllers;

import dev.jolvera.finjav.models.User;
import dev.jolvera.finjav.viewModels.AccountDialogViewModel;
import jakarta.inject.Inject;
import javafx.application.Platform;
import javafx.fxml.FXML;
import javafx.scene.control.Button;
import javafx.scene.control.Label;
import javafx.scene.control.PasswordField;
import javafx.scene.control.TextField;
import javafx.stage.Stage;

public class AccountDialogController {
    @FXML public TextField LoginUsernameInput;
    @FXML public PasswordField LoginPasswordInput;
    @FXML public Button LoginConfirmButton;
    @FXML public Button LoginClearFieldsButton;

    @FXML public TextField RegisterUsernameInput;
    @FXML public TextField RegisterEmailInput;
    @FXML public PasswordField RegisterPasswordInput;
    @FXML public Button RegisterConfirmButton;
    @FXML public Button RegisterClearFieldsButton;

    @FXML public Label errorLabel;

    private AccountDialogViewModel viewModel;
    private User result;

    @Inject
    public AccountDialogController(AccountDialogViewModel viewModel) {
        this.viewModel = viewModel;
    }

    @FXML
    private void initialize() {
        setupBindings();
        setupHandlers();
    }

    public User getResult() {
        return this.result;
    }

    private void setupBindings() {
        LoginUsernameInput.textProperty().bindBidirectional(viewModel.loginUsernameProperty());
        LoginPasswordInput.textProperty().bindBidirectional(viewModel.loginPasswordProperty());

        RegisterUsernameInput.textProperty().bindBidirectional(viewModel.registerUsernameProperty());
        RegisterEmailInput.textProperty().bindBidirectional(viewModel.registerEmailProperty());
        RegisterPasswordInput.textProperty().bindBidirectional(viewModel.registerPasswordProperty());

        errorLabel.textProperty().bindBidirectional(viewModel.errorMessageProperty());
        errorLabel.visibleProperty().bindBidirectional(viewModel.hasErrorProperty());

        // Bind button states
        LoginConfirmButton.disableProperty().bind(
                viewModel.isLoadingProperty().or(
                        LoginUsernameInput.textProperty().isEmpty().or(
                                LoginPasswordInput.textProperty().isEmpty()
                        )
                )
        );

        RegisterConfirmButton.disableProperty().bind(
                viewModel.isLoadingProperty().or(
                        RegisterUsernameInput.textProperty().isEmpty().or(
                                RegisterEmailInput.textProperty().isEmpty().or(
                                        RegisterPasswordInput.textProperty().isEmpty()
                                )
                        )
                )
        );

        viewModel.userProperty().addListener((observable, oldValue, newValue) -> {
            if (newValue != null) {
                this.result = newValue;

                Platform.runLater(() -> {
                    Stage stage = (Stage)LoginConfirmButton.getScene().getWindow();
                    stage.close();
                });
            }
        });
    }

    private void setupHandlers() {
        LoginConfirmButton.setOnAction(e -> handleLogin());
        LoginClearFieldsButton.setOnAction(e -> handleClearLoginFields());
        RegisterConfirmButton.setOnAction(e -> handleRegister());
        RegisterClearFieldsButton.setOnAction(e -> handleClearRegisterFields());
    }

    private void handleLogin() {
        viewModel.attemptLogin();
    }

    private void handleRegister() {
        viewModel.attemptRegister();
    }

    private void handleClearLoginFields() {
        LoginUsernameInput.clear();
        LoginPasswordInput.clear();
    }

    private void handleClearRegisterFields() {
        RegisterUsernameInput.clear();
        RegisterPasswordInput.clear();
        RegisterPasswordInput.clear();
    }

    private void closeDialog() {
        Stage stage = (Stage) LoginConfirmButton.getScene().getWindow();
        stage.close();
    }
}

File: src/main/java/dev/jolvera/finjav/controllers/AccountDialogController.kt
package dev.jolvera.finjav.controllers

import dev.jolvera.finjav.models.User
import dev.jolvera.finjav.viewModels.AccountDialogViewModel
import javafx.application.Platform
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.fxml.FXML
import javafx.scene.control.Button
import javafx.scene.control.Label
import javafx.scene.control.PasswordField
import javafx.scene.control.TextField
import javafx.stage.Stage
import org.koin.core.component.KoinComponent
import org.koin.core.component.get

class AccountDialogController: KoinComponent {
    private lateinit var viewModel: AccountDialogViewModel

    @FXML lateinit var loginNameInput: TextField
    @FXML lateinit var loginPasswordInput: PasswordField
    @FXML lateinit var loginConfirmButton: Button
    @FXML lateinit var loginClearFieldsButton: Button

    @FXML lateinit var registerNameInput: TextField
    @FXML lateinit var registerEmailInput: TextField
    @FXML lateinit var registerPasswordInput: PasswordField
    @FXML lateinit var registerConfirmButton: Button
    @FXML lateinit var registerClearFieldsButton: Button

    @FXML lateinit var errorLabel: Label

    private lateinit var result: User
    fun getResult(): User { return this.result }

    @FXML
    private fun initialize() {
        viewModel = get<AccountDialogViewModel>()
        setupBindings()
        setupHandlers()
    }

    private fun setupBindings() {
        loginNameInput.textProperty().bindBidirectional(viewModel.loginNameProperty())
        loginPasswordInput.textProperty().bindBidirectional(viewModel.loginPasswordProperty())

        registerNameInput.textProperty().bindBidirectional(viewModel.registerNameProperty())
        registerEmailInput.textProperty().bindBidirectional(viewModel.registerEmailProperty())
        registerPasswordInput.textProperty().bindBidirectional(viewModel.registerPasswordProperty())

        errorLabel.textProperty().bindBidirectional(viewModel.ErrorProperty())
        errorLabel.visibleProperty().bindBidirectional(viewModel.hasErrorProperty())

        loginConfirmButton.disableProperty().bind(
            viewModel.IsLoadingProperty().or(
                loginNameInput.textProperty().isEmpty.or(
                    loginPasswordInput.textProperty().isEmpty
                )
            )
        )

        registerConfirmButton.disableProperty().bind(
            viewModel.IsLoadingProperty().or(
                registerNameInput.textProperty().isEmpty().or(
                    registerEmailInput.textProperty().isEmpty().or(
                        registerPasswordInput.textProperty().isEmpty()
                    )
                )
            )
        )

        viewModel.returnedUserProperty().addListener { _, _, newValue ->
            if (newValue != null) {
                this.result = newValue

                Platform.runLater {
                    val stage = loginConfirmButton.scene.window as Stage
                    stage.close()
                }
            }
        }
    }

    private fun setupHandlers() {
        loginConfirmButton.onAction = EventHandler { e: ActionEvent? -> handleLogin() }
        loginClearFieldsButton.onAction = EventHandler { e: ActionEvent? -> handleClearLoginFields() }
        registerConfirmButton.onAction = EventHandler { e: ActionEvent? -> handleRegister() }
        registerClearFieldsButton.onAction = EventHandler { e: ActionEvent? -> handleClearRegisterFields() }
    }

    private fun handleLogin() { viewModel.attemptLogin()}
    private fun handleRegister() { viewModel.attemptRegister() }

    private fun handleClearLoginFields() {
        loginNameInput.clear()
        loginPasswordInput.clear()
    }
    private fun handleClearRegisterFields() {
        registerNameInput.clear()
        registerEmailInput.clear()
        registerPasswordInput.clear()
    }
}